### PR TITLE
Populate triggerKind in completeRequest.

### PR DESCRIPTION
**Pull request by tigersoldier**

### CODE REVIEW

**Comment by tigersoldier** [comment]
@MaskRay Mind trying this and see if it works for ccls?

@dgutov Any concern on the assumptions of `company-point` or how it's used in this PR?

@yyoncho With this one lsp-mode should register `ClientCapabilities.textDocument.completion.contextSupport` capability.

**Comment by yyoncho** [comment]
> @yyoncho With this one lsp-mode should register ClientCapabilities.textDocument.completion.contextSupport capability.

Sure, I will do the necessary adjustments. Will this address the caching issues you were talking about in the one of the previous PRs.

**Comment by tigersoldier** [comment]
> Will this address the caching issues you were talking about in the one of the previous PRs.

This is irrelevant. I believe ccls needs this to perform better in some cases.

File: company-lsp.el [hunk 1]
```diff
@@ -511,6 +543,56 @@ CALLBACK is a function that takes a list of strings as completion candidates."
...
+                       params)))
+
+(defun company-lsp--get-trigger-kind (prefix)
+  "Determine why completion is triggered for PREFIX.
+
+Returns one of `company-lsp--trigger-kind-invoked',
+`company-lsp--trigger-kind-trigger-character' and
+`company-lsp--trigger-kind-incomplete'."
+  (cond
+   ((or (eq this-command 'company-lsp)
```
Comment (dgutov) [inline company-lsp.el]: Shouldn't this include `company-complete` and `company-complete-common`, at least?

File: company-lsp.el [hunk 2]
```diff
@@ -511,6 +543,56 @@ CALLBACK is a function that takes a list of strings as completion candidates."
...
+        (eq this-command 'company-begin-backend))
+    ;; Explicitly calling completion command.
+    company-lsp--trigger-kind-invoked)
+   ((company-lsp--cache-find-closest prefix)
+    ;; Has incomplete candidates for sub-prefix. This assumes that if the
+    ;; candidates for sub-prefix, completion won't reach this step since the
+    ;; candidates for the sub-prefix can be narrowed down and returned directly.
+    company-lsp--trigger-kind-incomplete)
+   ((or (null company-point) (< company-point (point)))
```
Comment (dgutov) [inline company-lsp.el]: This is probably an okay approach, at least for the near future.

Eglot's solution seems more future-proof, though: https://github.com/joaotavora/eglot/blob/7d6e3cf5d7ae098aa6c8572343c8bc9b8453aace/eglot.el#L1504-L1510

File: company-lsp.el (hunk 1 above)
Comment (tigersoldier) [reply company-lsp.el]: Added

File: company-lsp.el (hunk 2 above)
Comment (tigersoldier) [reply company-lsp.el]: The reason why I prefer this one is that there is no way for company-lsp to remove its hooks. There is no `lsp-mode-disable-hook` or `company-mode-disable-hook` for it and `company-lsp` itself is not a minor hook.

**Pull request merged**